src/libopensc/card-sc-hsm.c

```diff
@@ -875,7 +875,7 @@ static int sc_hsm_write_ef(sc_card_t *card,
 		}
 
 		if (buf != NULL)
-			memcpy(p, buf+offset, to_send);
+			memcpy(p, buf, to_send);
```

Comment:
This looks like this is a general misconception about the `idx` parameter of this function. This should refer to the **index of the file** of the card we want to write to rather than to the index of the buffer we are supplying.

@istr could please you have a look in detail whether this has any impact of the rest of the code that was added in #2682?

Reply:
sc_update_binary is meant to write a buffer to a card's file with a specific offset, just like `man 2 pwrite`. Hence, its second parameter which is sometimes named `offs` is meant to specify the offset of the card's file. Please note that at the end of the `while` loop in `sc_update_binary` both, `idx` and `buf` are incremented by the number of bytes written. This means, that the card drivers' `update_binary` (i.e. `sc_hsm_update_binary`) should write `buf` at the file's offset `idx`.

Reply:
@frankmorgner 
> This means, that the card drivers' update_binary (i.e. sc_hsm_update_binary) should write buf at the file's offset idx.

Ok, I see. Then this needs to be fixed. The two offsets into `buf` (needs to start at `0`) and into the card's file (needs to start at `idx`) need to be tracked independently in this inner loop as well. I will provide a PR for that.

@Jakuje After this has been fixed, it should be sufficient to rely on the buffer limit check that is already performed in `sc_update_binary`.

Reply:
OK, then it would be best to remove this change from this PR and give you the time to look into this for a proper fix. thank you.

Reply:
Thank you for looking into this, I will remove this commit.